src/app/EventManagement.cpp

```diff
@@ -0,0 +1,903 @@
+ */
+class CircularEventReader : public chip::TLV::TLVReader
+{
+public:
+    /**
+     * @brief
+     *   Initializes a TLVReader object backed by CircularEventBuffer
+     *
+     * Reading begins in the CircularTLVBuffer belonging to this
+     * CircularEventBuffer.  When the reader runs out of data, it begins
+     * to read from the previous CircularEventBuffer.
+     *
+     * @param[in] apBuf A pointer to a fully initialized CircularEventBuffer
+     *
+     */
+    void Init(CircularEventBufferWrapper * apBuf);
+
+    virtual ~CircularEventReader() = default;
+};
+
+EventManagement & EventManagement::GetInstance(void)
+{
+    return sInstance;
+}
+
+struct ReclaimEventCtx
+{
+    CircularEventBuffer * mpEventBuffer = nullptr;
+    size_t mSpaceNeededForEvent         = 0;
+};
+
+/**
+ * @brief
+ *  Internal structure for traversing event list.
+ */
+struct CopyAndAdjustDeltaTimeContext
+{
+    CopyAndAdjustDeltaTimeContext(TLVWriter * aWriter, EventLoadOutContext * inContext) : mpWriter(aWriter), mpContext(inContext) {}
+
+    chip::TLV::TLVWriter * mpWriter = nullptr;
+    EventLoadOutContext * mpContext = nullptr;
+};
+
+/**
+ * @brief
+ *  Internal structure for traversing events.
+ */
+struct EventEnvelopeContext
+{
+    EventEnvelopeContext() {}
+
+    size_t mNumFieldsToRead =
+        2 /* PriorityLevel and DeltaSystemTimestamp are there if that is not first event when putting events in report*/;
+    uint32_t mDeltaTime     = 0;
+    int64_t mDeltaUtc       = 0;
+    PriorityLevel mPriority = PriorityLevel::First;
+};
+
+EventManagement::EventManagement(Messaging::ExchangeManager * apExchangeMgr, int aNumBuffers,
+                                 CircularEventBuffer * apCircularEventBuffer,
+                                 const LogStorageResources * const apLogStorageResources)
+{
+    Init(apExchangeMgr, aNumBuffers, apCircularEventBuffer, apLogStorageResources);
+}
+
+void EventManagement::Init(Messaging::ExchangeManager * apExchangeManager, int aNumBuffers,
+                           CircularEventBuffer * apCircularEventBuffer, const LogStorageResources * const apLogStorageResources)
+{
+    VerifyOrDie(aNumBuffers > 0);
+
+    mpExchangeMgr = apExchangeManager;
+
+    for (int bufferIndex = 0; bufferIndex < aNumBuffers; bufferIndex++)
+    {
+        static CircularEventBuffer * current = nullptr;
+        static CircularEventBuffer * prev    = nullptr;
+        static CircularEventBuffer * next    = nullptr;
+
+        next = (bufferIndex < aNumBuffers - 1) ? &apCircularEventBuffer[bufferIndex + 1] : nullptr;
+
+        current = &apCircularEventBuffer[bufferIndex];
+        current->Init(static_cast<uint8_t *>(apLogStorageResources[bufferIndex].mpBuffer),
+                      (uint32_t)(apLogStorageResources[bufferIndex].mBufferSize), prev, next);
+
+        prev = current;
+
+        current->mProcessEvictedElement = AlwaysFail;
+        current->mAppData               = nullptr;
```

Comment:
mAppData is used for ReclaimEventCtx, which has 
struct ReclaimEventCtx
{
    CircularEventBuffer * mpEventBuffer = nullptr;
    size_t mSpaceNeededForMovedEvent    = 0;
};

when debug buffer is full, we try to evict event to next buffer, we need to get to know how much space for evicted event in debug buffer 

EvictHead function(https://github.com/project-chip/connectedhomeip/blob/master/src/lib/core/CHIPCircularTLVBuffer.cpp#L117) would check and call this mProcessEvictedElement function, this function could be AlwaysFail or EvictEvent, if that is EvictEvent, it will check if the tail's event can be moved to higher priority, if not, dropped, if yes, note how much space it needs.

I would use delegate to get rid void pattern in follow-up PR, thanks

Reply:
https://github.com/project-chip/connectedhomeip/issues/6429